source/2_platform/rex_windows/src/gui_application.cpp

```diff
@@ -5,41 +5,186 @@
+  extern void rex_gl_swap_buffers()
+  {
+    SwapBuffers(s_glctx.dc);
+  }
+
+  namespace opengl
+  {
+    void release_device_context(HWND hwnd, HDC dc)
+    {
+      // The application must call the ReleaseDC function for each call to the GetWindowDC function
+      // and for each call to the GetDC function that retrieves a common DC.
+      // https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-releasedc
+      ReleaseDC(hwnd, dc);
+    }
+
+    bool create_gl_context(HWND hwnd)
+    {
+      s_glctx.dc = GetDC(hwnd);
+
+      PIXELFORMATDESCRIPTOR pfd;
+      memset(&pfd, 0, sizeof(pfd));
+
+      pfd.nSize = sizeof(PIXELFORMATDESCRIPTOR);
+      pfd.nVersion = 1;
+      pfd.dwFlags = PFD_DRAW_TO_WINDOW | PFD_SUPPORT_OPENGL | PFD_DOUBLEBUFFER;
+      pfd.iPixelType = PFD_TYPE_RGBA;
+      pfd.cColorBits = 32;
+      pfd.cAlphaBits = 8;
+      pfd.cDepthBits = 24;
+      pfd.cStencilBits = 8;
+      pfd.iLayerType = PFD_MAIN_PLANE;
+
+      s32 pf = ChoosePixelFormat(s_glctx.dc, &pfd);
+      if (pf == NULL || SetPixelFormat(s_glctx.dc, pf, &pfd) == FALSE)
+      {
+        release_device_context(hwnd, s_glctx.dc);
+        REX_ERROR("Failed to set a compatible pixel format");
+        return false;
+      }
+
+      HGLRC temp = wglCreateContext(s_glctx.dc);
+      if (temp == NULL)
+      {
+        release_device_context(hwnd, s_glctx.dc);
+        REX_ERROR("Failed to create the initial rendering context");
+        return false;
+      }
+
+      wglMakeCurrent(s_glctx.dc, temp);
+
+      // Load WGL Extensions:
+      gladLoaderLoadWGL(s_glctx.dc);
+
+      s32 attribs[] =
+      {
+        WGL_CONTEXT_MAJOR_VERSION_ARB, 3,       // Set the MAJOR version of OpenGL to 3
+        WGL_CONTEXT_MINOR_VERSION_ARB, 2,       // Set the MINOR version of OpenGL to 2
+        WGL_CONTEXT_FLAGS_ARB, 
+        WGL_CONTEXT_FORWARD_COMPATIBLE_BIT_ARB, // Set our OpenGL context to be forward compatible
+        0
+      };
+
+      const char* extensions = (const char*)wglGetExtensionsStringARB(s_glctx.dc);
+      if (strstr(extensions, "WGL_ARB_create_context") == NULL)
+      {
+        s_glctx.glrc = wglCreateContextAttribsARB(s_glctx.dc, NULL, attribs);
+        if (s_glctx.glrc == NULL)
+        {
+          wglDeleteContext(temp);
+          release_device_context(hwnd, s_glctx.dc);
+          REX_ERROR("Failed to create Windows OpenGL Context");
+
+          return false;
+        }
+
+        wglMakeCurrent(NULL, NULL);
+        wglDeleteContext(temp);
+        wglMakeCurrent(s_glctx.dc, s_glctx.glrc);
+      }
+      else
+      {
+        s_glctx.glrc = temp;
+      }
+
+      if (!gladLoaderLoadGL()) 
+      {
+        wglMakeCurrent(NULL, NULL);
+        wglDeleteContext(s_glctx.glrc);
+        release_device_context(hwnd, s_glctx.dc);
+        REX_ERROR("Glad Loader failed!");
+        
+        return false;
+      }
+
+      return true;
+    }
+  }
+}
+
+#define create_ctx(hwnd) rex::opengl::create_gl_context(hwnd)
```

Comment:
this can be a function and in the body you check which API you use and then call the right one

Reply:
So you would like me to create an enum with the available API's and switch on that when creating the context?